File: yandex_back_prep_cpp/20_Valid_Parentheses.cpp
#include <set>
#include <iostream>

using namespace std;

class Solution
{
public:
    bool isValid(string s)
    {
        std::vector<char> stack;

        for (const char c : s)
        {

            if (c == '(' | c == '{' | c == '[')
            {
                stack.push_back(c);
            }
            else
            {
                if (stack.size() == 0)
                {
                    return false;
                }

                char left = stack.back();
                stack.pop_back();

                std::string _s({left, c});
                if (!(_s == "{}" || _s == "[]" || _s == "()"))
                {
                    return false;
                }
            }
        }

        if (stack.size() == 0)
        {
            return true;
        }

        return false;
    }
};

int main()
{

    Solution s;
    bool res = s.isValid(std::string("(){}}"));

    std::cout << "res :" << res << std::endl;

    return 0;
}

File: yandex_back_prep_python/leetcode_python/20. Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        
        st = []
        p = ['()', '[]', '{}']

        for c in s:
            
            if c in ['{', '[', '(']:
                st.append(c)
            else:
                if not st:
                    return False

                _c = st.pop()
        
                if _c+c not in p:
                    return False

        return len(st) == 0



            